pkg/controller/controller.go

```diff
@@ -470,23 +470,70 @@ func getAccessMode(pvcAccessMode v1.PersistentVolumeAccessMode) *csi.VolumeCapab
 	}
 }
 
+func getSingleNodeMultiWriterCapableAccessMode(pvcAccessMode v1.PersistentVolumeAccessMode) *csi.VolumeCapability_AccessMode {
+	switch pvcAccessMode {
+	case v1.ReadWriteOnce:
+		return &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+		}
+	case v1.ReadWriteMany:
+		return &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
+		}
+	case v1.ReadOnlyMany:
+		return &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY,
+		}
+	case v1.ReadWriteOncePod:
+		return &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER,
+		}
+	default:
+		return nil
+	}
+}
+
 func getVolumeCapability(
 	claim *v1.PersistentVolumeClaim,
 	sc *storagev1.StorageClass,
 	pvcAccessMode v1.PersistentVolumeAccessMode,
 	fsType string,
+	supportsSingleNodeMultiWriter bool,
 ) *csi.VolumeCapability {
+	var accessMode *csi.VolumeCapability_AccessMode
+	if supportsSingleNodeMultiWriter {
+		accessMode = getSingleNodeMultiWriterCapableAccessMode(pvcAccessMode)
+	} else {
+		accessMode = getAccessMode(pvcAccessMode)
+	}
+
 	if util.CheckPersistentVolumeClaimModeBlock(claim) {
 		return &csi.VolumeCapability{
 			AccessType: getAccessTypeBlock(),
-			AccessMode: getAccessMode(pvcAccessMode),
+			AccessMode: accessMode,
 		}
 	}
 	return &csi.VolumeCapability{
 		AccessType: getAccessTypeMount(fsType, sc.MountOptions),
-		AccessMode: getAccessMode(pvcAccessMode),
+		AccessMode: accessMode,
 	}
+}
 
+func (p *csiProvisioner) getVolumeCapabilities(
+	claim *v1.PersistentVolumeClaim,
+	sc *storagev1.StorageClass,
+	fsType string,
+) []*csi.VolumeCapability {
+	supportsSingleNodeMultiWriter := false
+	if p.controllerCapabilities[csi.ControllerServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER] {
```

Comment:
what if controller only has SINGLE_NODE_SINGLE_WRITER?

Reply:
The capability name does not make it entirely clear, but in the CSI spec we define that the `SINGLE_NODE_MULTI_WRITER` capability means the driver supports the `SINGLE_NODE_SINGLE_WRITER` and/or `SINGLE_NODE_MULTI_WRITER` access modes.

The reason the capability is named `SINGLE_NODE_MULTI_WRITER` is because this access mode is missing from the CSI spec. This applies to drivers that strictly followed the NodePublishVolume rules which made `SINGLE_NODE_WRITER` == `SINGLE_NODE_SINGLE_WRITER`).

Reply:
Indeed, the capabilities in CSI spec  and its indirect capability pass through is confusing a lot :) . Regardless, with the addition of explicit `SINGLE_NODE_SINGLE_WRITER and SINGLE_NODE_MULTI_WRITER` in recent spec, the situation is getting improved. `ReadWriteOncePod` is much needed one too.  :+1: .